File: bdp-devops/ambari/ambari-web/test/utils/updater_test.js
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

var App = require('app');
require('utils/updater');
describe('utils/updater', function () {
  beforeEach(function () {
    this.clock = sinon.useFakeTimers();
    sinon.stub(App.router, "get").returns('test');
    App.updater.run.restore();
    sinon.spy(App.updater, 'run');
    App.updater.immediateRun.restore();
    sinon.spy(App.updater, 'immediateRun');
  });

  var tests = {
    t1: {
      obj: Em.Object.create({
        method: sinon.spy(),
        isWorking: true
      }),
      m: 'method called once with default interval in 15 000 ms'
    },
    t2: {
      obj: Em.Object.create({
        method: function () {
        }
      }),
      m: 'should return false if key name is invalid or absent'
    },
    t3: {
      obj: Em.Object.create({
        method2: sinon.spy(),
        isWorking: true
      }),
      m: 'method should be called immediately'
    },
    t4: {
      obj: Em.Object.create({
        method3: sinon.spy(),
        isWorking: true
      }),
      m: 'method call should be ignored if `isWorking` set to false'
    },
    t5: {
      obj: Em.Object.create({
        method4: sinon.spy(),
        isWorking: true
      }),
      m: 'method call should be ignored if urlPattern is not matching router location'
    }
  };

  it(tests.t1.m, function () {
    App.updater.run(tests.t1.obj, 'method', 'isWorking');
    this.clock.tick(3600000 * 1.5);
    expect(tests.t1.obj.method.called).to.be.ok;
  });

  it(tests.t2.m, function () {
    var methodCall = App.updater.run(tests.t2.obj, 'method', 'isWorking');
    expect(methodCall).to.be.false;
  });

  it(tests.t3.m, function () {
    App.updater.run(tests.t3.obj, 'method2', 'isWorking');
    App.updater.immediateRun('method2');
    expect(tests.t3.obj.method2.called).to.be.ok;
  });

  it(tests.t4.m, function () {
    App.updater.run(tests.t4.obj, 'method3', 'isWorking');
    this.clock.tick(10000);
    tests.t4.obj.set('isWorking', false);
    this.clock.tick(5000);
    expect(tests.t4.obj.method3.called).to.be.false;
  });

  it(tests.t5.m, function () {
    App.updater.run(tests.t5.obj, 'method4', 'isWorking', 15000, 'pattern');
    this.clock.tick(15000);
    expect(tests.t5.obj.method4.called).to.be.false;
  });

  afterEach(function () {
    this.clock.restore();
    App.router.get.restore();
  });
});

File: k8s/koordinator/pkg/koordlet/resourceexecutor/updater_test.go
/*
Copyright 2022 The Koordinator Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package resourceexecutor

import (
	"path/filepath"
	"testing"

	"github.com/stretchr/testify/assert"

	sysutil "github.com/koordinator-sh/koordinator/pkg/koordlet/util/system"
)

func TestNewCommonCgroupUpdater(t *testing.T) {
	type fields struct {
		UseCgroupsV2 bool
	}
	type args struct {
		resourceType sysutil.ResourceType
		parentDir    string
		value        string
	}
	tests := []struct {
		name    string
		fields  fields
		args    args
		want    ResourceUpdater
		wantErr bool
	}{
		{
			name: "updater not found",
			args: args{
				resourceType: sysutil.ResourceType("UnknownResource"),
			},
			want:    nil,
			wantErr: true,
		},
		{
			name: "get cfs quota updater",
			args: args{
				resourceType: sysutil.CPUCFSQuotaName,
				parentDir:    "/kubepods.slice/kubepods.slice-podxxx",
				value:        "-1",
			},
			want: &CgroupResourceUpdater{
				file:       sysutil.CPUCFSQuota,
				parentDir:  "/kubepods.slice/kubepods.slice-podxxx",
				value:      "-1",
				updateFunc: CommonCgroupUpdateFunc,
			},
			wantErr: false,
		},
		{
			name: "get cpu max updater",
			fields: fields{
				UseCgroupsV2: true,
			},
			args: args{
				resourceType: sysutil.CPUCFSQuotaName,
				parentDir:    "/kubepods.slice/kubepods.slice-podxxx",
				value:        "-1",
			},
			want: &CgroupResourceUpdater{
				file:       sysutil.CPUCFSQuotaV2,
				parentDir:  "/kubepods.slice/kubepods.slice-podxxx",
				value:      "-1",
				updateFunc: CommonCgroupUpdateFunc,
			},
			wantErr: false,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			helper := sysutil.NewFileTestUtil(t)
			defer helper.Cleanup()
			helper.SetCgroupsV2(tt.fields.UseCgroupsV2)

			got, gotErr := NewCommonCgroupUpdater(tt.args.resourceType, tt.args.parentDir, tt.args.value)
			if !tt.wantErr {
				assert.NotNil(t, got)
				assert.Equal(t, tt.want.ResourceType(), got.ResourceType())
				assert.Equal(t, tt.want.Path(), got.Path())
				assert.Equal(t, tt.want.Value(), got.Value())
			}
			assert.Equal(t, tt.wantErr, gotErr != nil, gotErr)
		})
	}
}

func TestCgroupResourceUpdater_Update(t *testing.T) {
	type fields struct {
		UseCgroupsV2 bool
		initialValue string
	}
	type args struct {
		resourceType sysutil.ResourceType
		parentDir    string
		value        string
	}
	tests := []struct {
		name    string
		fields  fields
		args    args
		want    string
		wantErr bool
	}{
		{
			name: "update cfs quota",
			fields: fields{
				initialValue: "10000",
			},
			args: args{
				resourceType: sysutil.CPUCFSQuotaName,
				parentDir:    "/kubepods.slice/kubepods.slice-podxxx",
				value:        "-1",
			},
			want:    "-1",
			wantErr: false,
		},
		{
			name: "update memory.min",
			fields: fields{
				UseCgroupsV2: true,
				initialValue: "0",
			},
			args: args{
				resourceType: sysutil.MemoryMinName,
				parentDir:    "/kubepods.slice/kubepods.slice-podxxx",
				value:        "1048576",
			},
			want:    "1048576",
			wantErr: false,
		},
		{
			name: "update failed since file not exist",
			args: args{
				resourceType: sysutil.MemoryMinName,
				parentDir:    "/kubepods.slice/kubepods.slice-podxxx",
				value:        "1048576",
			},
			want:    "",
			wantErr: true,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			helper := sysutil.NewFileTestUtil(t)
			defer helper.Cleanup()
			helper.SetCgroupsV2(tt.fields.UseCgroupsV2)

			u, gotErr := NewCommonCgroupUpdater(tt.args.resourceType, tt.args.parentDir, tt.args.value)
			assert.NoError(t, gotErr)
			c, ok := u.(*CgroupResourceUpdater)
			assert.True(t, ok)
			if tt.fields.initialValue != "" {
				helper.WriteCgroupFileContents(tt.args.parentDir, c.file, tt.fields.initialValue)
			}

			gotErr = u.Update()
			assert.Equal(t, tt.wantErr, gotErr != nil)
			if !tt.wantErr {
				assert.Equal(t, tt.want, helper.ReadCgroupFileContents(c.parentDir, c.file))
			}
		})
	}
}

func TestDefaultResourceUpdater_Update(t *testing.T) {
	type fields struct {
		initialValue string
	}
	type args struct {
		file   string
		subDir string
		value  string
	}
	tests := []struct {
		name    string
		fields  fields
		args    args
		want    string
		wantErr bool
	}{
		{
			name: "update test file successfully",
			fields: fields{
				initialValue: "1234",
			},
			args: args{
				file:   "test_file",
				subDir: "test_dir",
				value:  "5678",
			},
			want:    "5678",
			wantErr: false,
		},
		{
			name: "update test file successfully 1",
			fields: fields{
				initialValue: "5678",
			},
			args: args{
				file:   "test_file",
				subDir: "test_dir",
				value:  "5678",
			},
			want:    "5678",
			wantErr: false,
		},
		{
			name: "update failed since file not exist",
			args: args{
				file:   "test_file1",
				subDir: "test_dir1",
				value:  "5678",
			},
			want:    "",
			wantErr: true,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			helper := sysutil.NewFileTestUtil(t)
			defer helper.Cleanup()

			file := filepath.Join(helper.TempDir, tt.args.subDir, tt.args.file)
			u, gotErr := NewCommonDefaultUpdater(file, file, tt.args.value)
			assert.NoError(t, gotErr)
			_, ok := u.(*DefaultResourceUpdater)
			assert.True(t, ok)
			if tt.fields.initialValue != "" {
				helper.WriteFileContents(filepath.Join(tt.args.subDir, tt.args.file), tt.fields.initialValue)
			}

			gotErr = u.Update()
			assert.Equal(t, tt.wantErr, gotErr != nil, gotErr)
			if !tt.wantErr {
				assert.Equal(t, tt.want, helper.ReadFileContents(filepath.Join(tt.args.subDir, tt.args.file)))
			}
		})
	}
}
